pkg/kv/store.go

```diff
@@ -0,0 +1,97 @@
+package kv
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"sync"
+)
+
+var (
+	ErrUnknownDriver   = errors.New("unknown driver")
+	ErrOperationFailed = errors.New("operation failed")
+	ErrMissingKey      = errors.New("missing key")
+	ErrMissingValue    = errors.New("missing value")
+	ErrNotFound        = errors.New("not found")
+	ErrClosedEntries   = errors.New("closed entries")
+)
+
+type Entry struct {
+	Key   []byte
+	Value []byte
+}
+
+type Driver interface {
+	// Open access to the database store. migrate kv if needed
+	Open(ctx context.Context, dsn string) (Store, error)
+
+	// Destroy will delete kv from storage
+	// Destroy(ctx context.Context, name string) error
+}
+
+type Store interface {
+	// Get returns a value for the given key, or ErrNotFound if key doesn't exist
+	Get(ctx context.Context, key []byte) ([]byte, error)
+
+	// Set stores the given value, overwriting an existing value if one exists
+	Set(ctx context.Context, key, value []byte) error
+
+	// SetIf returns an ErrPredicateFailed error if the valuePredicate passed
+	//  doesn't match the currently stored value. SetIf is a simple compare-and-swap operator:
+	//  valuePredicate is either the existing value, or an opaque value representing it (hash, index, etc).
```

Comment:
So how would the caller know if `valuePredicate` should be compared to the stored `value` or `hash`, `index`?

Reply:
Update the description - it should be the current value or nil for no key.